beep/helpers/featurizer_helpers.py

```diff
@@ -285,109 +275,218 @@ def get_hppc_ocv_helper(cycle_hppc_0, step_num):
+    """
+    This function narrows your data down to a dataframe that contains only the diagnostic cycle number you
+    are interested in.
+
+    Args:
+        processed_cycler_run (beep.structure.ProcessedCyclerRun)
+        diag_pos (int): diagnostic cycle occurence for a specific <diagnostic_cycle_type>. e.g.
+        if rpt_0.2C, occurs at cycle_index = [2, 42, 147, 249 ...], <diag_pos>=0 would correspond to cycle_index 2.
 
-def get_hppc_r(processed_cycler_run, diag_num):
-    '''
-    This function takes in cycling data for one cell and returns the resistance at different SOCs with resistance at the
-    first hppc cycle(cycle 2) deducted
-    Argument:
-            processed_cycler_run(process_cycler_run object)
-            diag_num(int): diagnostic cycle number at which you want to get the feature, such as 37 or 142
     Returns:
-            two floats
-            the variance of the diag_num - cycle 2 for HPPC resistance for both charge and discharge
-    '''
+        a datarame that only has the diagnostic cycle you are interested in, and there is a column called
+        'diagnostic_time[h]' starting from 0 for this dataframe.
+    """
+
     data = processed_cycler_run.diagnostic_interpolated
-    cycle_hppc = data.loc[data.cycle_type == 'hppc']
-    cycle_hppc = cycle_hppc.loc[cycle_hppc.current.notna()]
-    cycles = cycle_hppc.cycle_index.unique()
-    if diag_num not in cycles:
+    hppc_cycle = data.loc[data.cycle_type == 'hppc']
+    hppc_cycle = hppc_cycle.loc[hppc_cycle.current.notna()]
+    cycles = hppc_cycle.cycle_index.unique()
+    diag_num = cycles[diag_pos]
+
+    chosen = hppc_cycle.loc[hppc_cycle.cycle_index == diag_num]
+    chosen = chosen.sort_values(by='test_time')
+    chosen['diagnostic_time'] = (chosen.test_time - chosen.test_time.min()) / 3600
+
+    return chosen
+
+
+def res_calc(chosen, diag_pos, soc, step_ocv, step_cur, index):
+    """
+    This function calculates resistance based on different socs and differnet time scales in hppc cycles.
+
+    Args:
+        chosen(pd.DataFrame): a dataframe for a specific diagnostic cycle you are interested in.
+        diag_pos (int): diagnostic cycle occurence for a specific <diagnostic_cycle_type>. e.g.
+        if rpt_0.2C, occurs at cycle_index = [2, 42, 147, 249 ...], <diag_pos>=0 would correspond to cycle_index 2
+        soc (int): step index counter corresponding to the soc window of interest.
+        step_ocv (int): 0 corresponds to the 1h-rest, and 2 corresponds to the 40s-rest.
+        step_cur (int): 1 is for discharge, and 3 is for charge.
+        index (float or str): this will input a time scale for resistance (unit is second), e.g. 0.01, 5 or
+        'last' which is the entire pulse duration.
+
+    Returns:
+        (a number) resistance at a specific soc in hppc cycles
+    """
+
+    counters = []
+
+    if diag_pos == 0:
+        steps = [11, 12, 13, 14, 15]
+    else:
+        steps = [43, 44, 45, 46, 47]
```

Comment:
How are these steps set? Are they machine-generated? Curious as to why they're different for the first set of diagnostic cycles and the rest.

Reply:
Yes, I also find this very curious, the first cycle uses a different set of step numbers while later cycles are all the same

Reply:
This comes from the structure of the procedure file template. The first set of diagnostics and the first set of regular cycles are a different part of the procedure from the rest. This mean that they have different step numbers than the other diagnostic cycles.